Add First, Last, Pos, TotalItems and EvenOdd helpers

wilr opened the pull request:
When in an Element template the expected $First, $Last helpers aren't there so this replaces them with something.

robbieaverill commented on src/Models/BaseElement.php:
```diff
@@ -822,4 +830,28 @@ public function getPageTitle()
+     */
+    public function isFirstElement()
+    {
+        return !($this->Parent()->Elements()->filter('Sort:LessThan', $this->Sort)->exists());
```
What do you reckon about `$this->Parent()->Elements()->first() === $this`?

wilr replied:
Would `$this` be robust enough? I'm just thinking of objects perhaps changing at runtime with something like fluent etc.. 

Perhaps `$this->Parent()->Elements()->first()->ID === $this->ID`. ID is all that matters right?

robbieaverill replied:
Yeah give that a go. Your tests should be good enough to try it out =D

robbieaverill commented:
Related: #211

wilr commented:
@robbieaverill What to you think like #211 should we do something at the ElementArea level or leave it as functions on `BaseElement` if we leave it as methods on BaseElement then perhaps it would be more expected behaviour that `First`, `Last` be the method names rather than `isFirstElement`

robbieaverill commented:
@wilr I'm happy to follow your judgement =) they sound like useful methods either way

wilr commented:
Ok my judgement is consistently probably out weighs risk since the behaviour for all the options is going to be the same unless you start to do funky things like building your own `ArrayList` of them. In which case you could override First, Last anyway if they didn't work.

Updated the code, tests and docs. Changed branch to 3

robbieaverill commented on src/Models/BaseElement.php:
```diff
@@ -822,4 +830,48 @@ public function getPageTitle()
+    }
+
+    /**
+     * @return boolean
```
Minor: should be int

robbieaverill commented on src/Models/BaseElement.php:
```diff
@@ -822,4 +830,48 @@ public function getPageTitle()
+    /**
+     * Returns the position of the current element.
+     *
+     * @return boolean
```
int

robbieaverill commented on src/Models/BaseElement.php:
```diff
@@ -822,4 +830,48 @@ public function getPageTitle()
+    }
+
+    /**
+     * @return boolean
```
string

ScopeyNZ commented on src/Models/BaseElement.php:
```diff
@@ -239,7 +239,15 @@ public function onBeforeWrite()
+            if ($this->hasExtension(Versioned::class)) {
+                $records = Versioned::get_by_stage(BaseElement::class, Versioned::DRAFT);
+            } else {
+                $records = BaseElement::get()->max('Sort');
```
Sorry for the late review but it appears this is a bug?

wilr replied:
@ScopeyNZ What seems to be the bug? I added this in as it was skipping draft version numbers and it was also using the global max rather than the local max.

ScopeyNZ replied:
`$records = ...->max('Sort')` should return an int, but is treated like an object on the next line?

wilr replied:
Ah yep, but only if versioned had been removed with is rare. I'll submit a fix to the 3 branch.

Thanks

ScopeyNZ replied:
Sweet 👍
